piet-coregraphics/src/text.rs

```diff
@@ -0,0 +1,433 @@
+        // fall between the preceding baseline and that line's descent
+        if line_num > 0 {
+            let prev_line = self.unwrap_frame().get_line(line_num - 1).unwrap();
+            let typo_bounds = Line::new(&&prev_line).get_typographic_bounds();
+            if self.line_y_positions[line_num - 1] + typo_bounds.descent >= point.y {
+                line_num -= 1;
+            }
+        }
+        let line: Line = self
+            .unwrap_frame()
+            .get_line(line_num)
+            .map(Into::into)
+            .unwrap();
+        let fake_y = self.line_y_positions[line_num];
+        // map that back into our inverted coordinate space
+        let fake_y = -(self.frame_size.height - fake_y);
+        let point_in_string_space = CGPoint::new(point.x, fake_y);
+        let offset_utf16 = line.get_string_index_for_position(point_in_string_space);
+        let offset = match offset_utf16 {
+            // this is 'kCFNotFound'.
+            // if nothing is found just go end of string? should this be len - 1? do we have an
+            // implicit newline at end of file? so many mysteries
+            -1 => self.string.len(),
+            n if n >= 0 => {
+                let utf16_range = line.get_string_range();
+                let utf8_range = self.line_range(line_num).unwrap();
+                let line_txt = self.line_text(line_num).unwrap();
+                let rel_offset = (n - utf16_range.location) as usize;
+                let mut off16 = 0;
+                let mut off8 = 0;
+                for c in line_txt.chars() {
+                    if rel_offset == off16 {
+                        break;
+                    }
+                    off16 += c.len_utf16();
+                    off8 += c.len_utf8();
+                }
+                utf8_range.0 + off8
+            }
+            // some other value; should never happen
+            _ => panic!("gross violation of api contract"),
+        };
+
+        let typo_bounds = line.get_typographic_bounds();
+        let is_inside_y = point.y >= 0. && point.y <= self.frame_size.height;
+        let is_inside_x = point.x >= 0. && point.x <= typo_bounds.width;
+
+        HitTestPoint {
+            metrics: HitTestMetrics {
+                text_position: offset,
+            },
+            is_inside: is_inside_x && is_inside_y,
+        }
+    }
+
+    fn hit_test_text_position(&self, offset: usize) -> Option<HitTestTextPosition> {
+        let line_num = match self.line_offsets.binary_search(&offset) {
+            Ok(line) => line.saturating_sub(1),
+            Err(line) => line.saturating_sub(1),
+        };
+        let line: Line = self.unwrap_frame().get_line(line_num)?.into();
+        let text = self.line_text(line_num)?;
+
+        let offset_remainder = offset - self.line_offsets.get(line_num)?;
+        let off16: usize = text[..offset_remainder].chars().map(char::len_utf16).sum();
```

Comment:
This is is a clear statement of what it's trying to compute but we also have a slightly more optimized helper function (nearby link above).

Reply:
can fix this when #184 lands.